learning_resources/serializers.py

```diff
@@ -754,38 +754,61 @@ class ContentFileSerializer(serializers.ModelSerializer):
-        source="run.learning_resource.topics", many=True
-    )
-    resource_id = serializers.CharField(source="run.learning_resource.id")
-    departments = LearningResourceDepartmentSerializer(
-        source="run.learning_resource.departments", many=True
-    )
-    resource_readable_id = serializers.CharField(
-        source="run.learning_resource.readable_id"
-    )
+    run_id = serializers.IntegerField(source="run.id", required=False)
+    run_readable_id = serializers.CharField(source="run.run_id", required=False)
+    run_title = serializers.CharField(source="run.title", required=False)
+    run_slug = serializers.CharField(source="run.slug", required=False)
+    semester = serializers.CharField(source="run.semester", required=False)
+    year = serializers.IntegerField(source="run.year", required=False)
+    topics = serializers.SerializerMethodField()
+    resource_id = serializers.SerializerMethodField()
+    departments = serializers.SerializerMethodField()
+    resource_readable_id = serializers.SerializerMethodField()
     course_number = serializers.SerializerMethodField()
     content_feature_type = LearningResourceContentTagField(source="content_tags")
-    offered_by = LearningResourceOfferorSerializer(
-        source="run.learning_resource.offered_by"
-    )
-    platform = LearningResourcePlatformSerializer(
-        source="run.learning_resource.platform"
-    )
+    offered_by = serializers.SerializerMethodField()
+    platform = serializers.SerializerMethodField()
+
+    def get_learning_resource(self, instance):
+        if instance.run:
+            return instance.run.learning_resource
+        return instance.learning_resource
+
+    @extend_schema_field(LearningResourcePlatformSerializer())
+    def get_platform(self, instance):
+        platform = self.get_learning_resource(instance).platform
+        return LearningResourcePlatformSerializer(platform).data
+
+    @extend_schema_field(LearningResourceOfferorSerializer())
+    def get_offered_by(self, instance):
+        offered_by = self.get_learning_resource(instance).offered_by
+        return LearningResourceOfferorSerializer(offered_by).data
+
+    @extend_schema_field({"type": "string"})
+    def get_resource_readable_id(self, instance):
+        return self.get_learning_resource(instance).readable_id
+
+    @extend_schema_field(LearningResourceDepartmentSerializer(many=True))
+    def get_departments(self, instance):
+        return LearningResourceDepartmentSerializer(
+            self.get_learning_resource(instance).departments, many=True
+        ).data
+
+    @extend_schema_field({"type": "string"})
+    def get_resource_id(self, instance):
+        return str(self.get_learning_resource(instance).id)
+
+    @extend_schema_field(LearningResourceTopicSerializer(many=True))
+    def get_topics(self, instance):
+        return LearningResourceTopicSerializer(
+            self.get_learning_resource(instance).topics, many=True
+        ).data
```

Comment:
Getting lots of n+1 queries now when querying the opensearch contentfiles (and probably when indexing them too), probably due to the above changes, so needs to be refactored.

Reply:
Resolved the n+1 queries